// google-assistant-node/lib/audio-transcriber.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var constants_1 = require("./constants");
var GCloudSpeech = require('@google-cloud/speech');
var AudioTranscriber = (function () {
    function AudioTranscriber(cloudProjectId) {
        this.speechClient = GCloudSpeech();
    }
    AudioTranscriber.prototype.createTranscriptionStream = function (options) {
        var encoding = constants_1.Encoding[options.encoding];
        console.log(options);
        var request = {
            config: {
                encoding: encoding,
                sampleRateHertz: options.sampleRateHertz,
                languageCode: options.languageCode
            },
            verbose: true
        };
        return this.speechClient.createRecognizeStream(request)
            .on('error', console.error);
    };
    AudioTranscriber.prototype._strToEncoding = function (str) {
        return constants_1.Encoding[str];
    };
    return AudioTranscriber;
}());
exports.default = AudioTranscriber;
